src/mp4.zig

```diff
@@ -0,0 +1,212 @@
+    // respects the following layout which seem to be standard:
+    //
+    // moov
+    //   udta
+    //     meta
+    //       ilst
+    //         aART
+    //         \xA9alb
+    //         \xA9ART
+    //
+    // The data that interests us are the atoms under the "ilst" atom.
+    //
+    // The following parser code expects this layout and if it doesn't exist it just fails.
+
+    var state: enum {
+        start,
+        in_moov,
+        in_udta,
+        in_meta,
+        in_ilst,
+    } = .start;
+
+    // Keep track of the size in bytes of the "ilst" atom and how much we already read.
+    var ilst_size: usize = 0;
+    var ilst_read: usize = 0;
+
+    while (true) {
+        const atom_header = readAtomHeader(reader, seekable_stream) catch |err| switch (err) {
+            error.EndOfStream => if (metadata.map.entries.items.len > 0) return metadata else return err,
+            else => return err,
+        };
+
+        switch (state) {
+            .start => if (std.mem.eql(u8, "moov", &atom_header.name)) {
+                state = .in_moov;
+                continue;
+            },
+            .in_moov => if (std.mem.eql(u8, "udta", &atom_header.name)) {
+                state = .in_udta;
+                continue;
+            },
+            .in_udta => if (std.mem.eql(u8, "meta", &atom_header.name)) {
+                // The full atom header doesn't interest us but it has to be read.
+                _ = try readFullAtomHeader(reader);
+
+                // The "meta" atom started at the current stream position minus the standard and full atom header.
+                metadata.start_offset = (try seekable_stream.getPos()) - 8 - 4;
+                metadata.end_offset = metadata.start_offset + atom_header.size;
+
+                state = .in_meta;
+                continue;
+            },
+            .in_meta => if (std.mem.eql(u8, "ilst", &atom_header.name)) {
+                // Used when handling the in_ilst state to know if there are more elements in the list.
+                ilst_size = atom_header.size - 8;
+
+                state = .in_ilst;
+                continue;
+            },
+            .in_ilst => {
+                // Determine if there's more to read in the "ilst" atom.
+                ilst_read += atom_header.size;
+                if (ilst_read >= ilst_size) {
+                    ilst_read = 0;
+                    ilst_size = 0;
+                    state = .start;
+                }
+
+                if (getMetadataAtom(&atom_header.name)) |atom| {
+                    const data_atom = try readDataAtom(allocator, reader, seekable_stream);
+                    defer allocator.free(data_atom.value);
+
+                    try metadata.map.put(atom.export_name, data_atom.value);
+
+                    continue;
+                }
+            },
+        }
+
+        // Skip every atom we don't recognize or are not interested in.
+
+        try seekable_stream.seekBy(atom_header.size - 8);
```

Comment:
This can overflow if the header size is set as less than 8 (an invalid header size). Adding

```zig
    if (size < 8) {
        return error.AtomSizeTooSmall;
    }
```
after the `const size = switch (try reader.readIntBig(u32)) {}` in  `readAtomHeader` would be one possible fix.